src/main/java/org/apache/commons/lang3/ArrayUtils.java

```diff
@@ -0,0 +1,9788 @@
+     * ArrayUtils.removeElements([], 1, 2)        = []
+     * ArrayUtils.removeElements([1], 2, 3)       = [1]
+     * ArrayUtils.removeElements([1, 3], 1, 2)    = [3]
+     * ArrayUtils.removeElements([1, 3, 1], 1)    = [3, 1]
+     * ArrayUtils.removeElements([1, 3, 1], 1, 1) = [3]
+     * </pre>
+     *
+     * @param array  the array to remove the element from, may be {@code null}
+     * @param values the elements to be removed
+     * @return A new array containing the existing elements except the
+     *         earliest-encountered occurrences of the specified elements.
+     * @since 3.0.1
+     */
+    public static int[] removeElements(final int[] array, final int... values) {
+        if (isEmpty(array) || isEmpty(values)) {
+            return clone(array);
+        }
+        final HashMap<Integer, MutableInt> occurrences = new HashMap<>(values.length);
+        for (final int v : values) {
+            final Integer boxed = Integer.valueOf(v);
+            final MutableInt count = occurrences.get(boxed);
+            if (count == null) {
+                occurrences.put(boxed, new MutableInt(1));
+            } else {
+                count.increment();
+            }
+        }
+        final BitSet toRemove = new BitSet();
+        for (int i = 0; i < array.length; i++) {
+            final int key = array[i];
+            final MutableInt count = occurrences.get(key);
+            if (count != null) {
+                if (count.decrementAndGet() == 0) {
+                    occurrences.remove(key);
+                }
+                toRemove.set(i);
+            }
+        }
+        return (int[]) removeAll(array, toRemove);
+    }
+
+    /**
+     * Removes occurrences of specified elements, in specified quantities,
+     * from the specified array. All subsequent elements are shifted left.
+     * For any element-to-be-removed specified in greater quantities than
+     * contained in the original array, no change occurs beyond the
+     * removal of the existing matching items.
+     * <p>
+     * This method returns a new array with the same elements of the input
+     * array except for the earliest-encountered occurrences of the specified
+     * elements. The component type of the returned array is always the same
+     * as that of the input array.
+     * </p>
+     * <pre>
+     * ArrayUtils.removeElements(null, 1, 2)      = null
+     * ArrayUtils.removeElements([], 1, 2)        = []
+     * ArrayUtils.removeElements([1], 2, 3)       = [1]
+     * ArrayUtils.removeElements([1, 3], 1, 2)    = [3]
+     * ArrayUtils.removeElements([1, 3, 1], 1)    = [3, 1]
+     * ArrayUtils.removeElements([1, 3, 1], 1, 1) = [3]
+     * </pre>
+     *
+     * @param array  the array to remove the element from, may be {@code null}
+     * @param values the elements to be removed
+     * @return A new array containing the existing elements except the
+     *         earliest-encountered occurrences
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;valueForNull&#39; (index: 1 | type: byte). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;array&#39;. The same parameter type is validated here for example: src/main/java/org/apache/commons/lang3/ArrayUtils.java:1549. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).